trunk/src/meshmod/refresh.F90

```diff
@@ -108,5 +107,92 @@ subroutine refresh
    NRDOFSE = NRDOFSE + nrdofE
    NRDOFSV = NRDOFSV + nrdofV
    NRDOFSQ = NRDOFSQ + nrdofQ
+!
+   if ((.not. DISTRIBUTED) .or. HOST_MESH) goto 99
+!
+!-----------------------------------------------------------------------
+!  Remark: We must make sure that unconstrained active nodes are only
+!          marked by a subdomain if they are part of a modified element
+!          of the subdomain.
+!          Otherwise, this creates issues with determining node
+!          ownership in solver interfaces without having to rerun
+!          logic_nodes to obtain modified element nodes.
+!          Without this fix, DOFs were also stored unnecessarily for
+!          some active nodes outside of the subdomain.
+!
+!  An example where an edge was marked incorrectly:
+!  (the same issue can happen for a face node)
+!
+!      The unconstrained, active       In href, edge0's kref already
+!      edge0 is marked by subd 0       coincides with the desired href
+!      b/c it is part of one of        but "break" always calls
+!      its modified element lists;     "activate_sons" which sets subd
+!      edge1 is constrained and        values for edge0's sons, incl.
+!      only marked by subd 1 as        edge1, based on edge0's subd.
+!      one of its local elem nodes.    So edge1 is marked by subd 0,
+!                                      even though it is neither on
+!                                      any modified element nor local
+!                                      element node list of subd 0.
+!
+!                edge0                           edge0
+!                  |            href ->            |
+!                  |                               |
+!                  v                               v
+!      *-----------*-----------*       *-----------*-----------*
+!      |           |           |       |           |           |
+!      |  subd 0   |           |       |  subd 0   |  subd 1   |
+!      |           |           |       |           |           |
+!      *-----------*  subd 1   |       *-----------*-----------*
+!      |           |           |       |           |           |
+!      |  subd 1   | <-edge1   |       |  subd 1   |  subd 1   |
+!      |           |           |       |           | <-edge1   |
+!      *-----------*-----------*       *-----------*-----------*
+!
+!-----------------------------------------------------------------------
+!
+!$OMP PARALLEL PRIVATE(iel,mdle,nod,subd)
+!
+!..1. Reset subdomain values
+!$OMP DO
+   do nod=1,NRNODS
+      call get_subd(nod, subd)
+      if ((subd.eq.RANK) .and. (.not.Is_middle(nod))) then
+         call set_subd(nod,-1)
+      endif
+   enddo
+!$OMP END DO
+!
+!..2. Set subdomains for all (unconstrained) nodes within subdomain
+!$OMP DO
+   do iel=1,NRELES_SUBD
+      mdle = ELEM_SUBD(iel)
+      call set_subd_elem(mdle)
```

Comment:
I'm not sure I love this; I like that it only calls `elem_nodes` on the subdomain (so is scalable) but it does add another call to `elem_nodes` to this routine. With the type changes it should be faster but this routine is called up to 6-7 times per grid level on large adaptive problems. This routine (and close_mesh) was a major bottleneck for large adaptive problems. 

Here's my version of Step 1 (I have similar changes for close_mesh). It relies on my bitvisit module (that's where call visit() and reduce_visit() come from); I'll work on committing that once I have my code working with the node type changes. Since we already mark all the nodes connected to a rank's subdomain in step 1 we could simply loop over NODES and check if there are any active nodes (with subd set?) that have not been visited.

It's probably fine for now; I can work on integrating these changes along with the other ones.

```suggestion 
!
!--------------------------------------------------------------------
! Step 1 : raise visitation flag for vertex, edge and face nodes of |
!          all active elements                                      |
!--------------------------------------------------------------------
!$omp parallel default(shared)
!$omp do private(iel,mdle,subd,type,i,nodesl,norientl) schedule(guided)
   do iel=1,NRELES_SUBD
!
      mdle = ELEM_SUBD(iel)
!
      call elem_nodes(mdle, nodesl,norientl)
!
      ntype=NODES(mdle)%ntype
      nvef = nvert(ntype)+nedge(ntype)+nface(ntype)
!
      do i=1,nvef
!     ...bitflag stores visit in bit collection (better for MPI)
         call visit(nodesl(i))
!
!     ...if node is visited by an element within my subdomain,
!        add node to my subdomain (need its dofs). this flag will
!        indicate that dofs must be allocated in activation.
         if (DISTRIBUTED) then
            call set_subd(nodesl(i),subd)
         endif
      enddo
!
   enddo
!$omp end do
!$omp end parallel
!
!..Reduce over node flag
   if (DISTRIBUTED) then
      call reduce_visit
   endif
```

Reply:
We can definitely optimize the visitation of nodes as needed if your bitvisit module is faster. The suggested change here would need to be adapted though--it's missing constrained nodes.

Reply:
I see; this is probably fine then, to get constrained nodes we'll need another call anyway.

Bitvisit just marks visitation in an array that can be quickly and conveniently MPI_reduced. When I bit encode the visitation flags I have to put atomic directives and it turns out that can be fairly slow with a lot of threads so for now I'm actually not bitencoding, just marking with integers and reducing.